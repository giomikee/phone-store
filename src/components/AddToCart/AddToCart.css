.add-to-cart .add-to-cart__price {
    margin-top: 0;
}