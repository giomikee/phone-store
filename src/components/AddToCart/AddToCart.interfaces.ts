export interface Props {
    onAddToCart: () => void;
    brand?: string;
    name?: string;
    disabled?: boolean;
    price?: number;
}