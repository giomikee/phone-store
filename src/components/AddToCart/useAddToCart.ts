import { useState } from 'react';

export function useAddToCart(onAddToCart: () => void) {
    const [isSnackbarOpen, setIsSnackbarOpen] = useState(false);

    const addToCart = () => {
        onAddToCart();
        setIsSnackbarOpen(true);
    };


    return {
        isSnackbarOpen,
        setIsSnackbarOpen,
        addToCart,
    };
}