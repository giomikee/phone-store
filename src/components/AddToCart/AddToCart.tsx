import { Alert, Button, Snackbar, Stack, Typography } from '@mui/material';
import AddShoppingCartIcon from '@mui/icons-material/AddShoppingCart';
import type { Props } from './AddToCart.interfaces';
import { formatPrice } from '../../utils';
import { useAddToCart } from './useAddToCart';
import './AddToCart.css';

const AddToCart = ({ disabled, brand, name, price, onAddToCart }: Props) => {
    const { addToCart, isSnackbarOpen, setIsSnackbarOpen }= useAddToCart(onAddToCart);


    return (
        <Stack 
            className='add-to-cart'
            spacing={2}
            gap={1}
            alignItems='center'
            sx={{
                flexDirection: {
                    xs: 'column-reverse',
                    md: 'row',
                }
            }}
        >
            <Button
                color='success'
                data-testid='add-to-cart-button'
                disabled={disabled}
                onClick={addToCart}
                size='large'
                startIcon={<AddShoppingCartIcon />}
                variant='contained'
                sx={{
                    width: {
                        xs: '100%',
                        md: 'auto'
                    }
                }}
            >
                Añadir al carrito
            </Button>

            {
                price &&
                    <Typography
                        className='add-to-cart__price'
                        color='primary'
                        component='h3'
                        data-testid='add-to-cart-price'
                        variant='h5'
                    >
                        {formatPrice(price)}
                    </Typography>
            }

            <Snackbar 
                data-testid='add-to-cart-snackbar'
                open={isSnackbarOpen}
                onClose={() => setIsSnackbarOpen(false)}
                autoHideDuration={2000}
            >
                <Alert
                    severity='success'
                    variant='filled'
                >
                    {brand} {name} añadido al carrito
                </Alert>
            </Snackbar>
        </Stack>
    )
}

export default AddToCart