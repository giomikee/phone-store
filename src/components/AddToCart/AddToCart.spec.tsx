
import { describe, expect, it, vi, beforeAll, afterAll } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import AddToCart from './AddToCart';

describe('Given AddToCart component', () => {
    const defaultProps = {
        disabled: false,
        brand: 'Apple',
        name: 'iPhone 15',
        price: 999,
        onAddToCart: vi.fn(),
    };

    describe('When the component is rendered', () => {
        it('Then it should match the snapshot', () => {
            const { container } = render(<AddToCart {...defaultProps} />);
            expect(container).toMatchSnapshot();
        })
    });

    [true, false].forEach((disabled) => {
        describe(`When the component is rendered with disabled=${disabled}`, () => {
            it(`Then add to cart button should be ${disabled ? 'disabled' : 'enabled'}`, () => {
                render(<AddToCart {...defaultProps} disabled={disabled} />);
                const button = screen.getByTestId('add-to-cart-button');
                expect(button).toBeInTheDocument();

                if (disabled) {
                    expect(button).toBeDisabled();
                } else {
                    expect(button).toBeEnabled();
                }
            })

            if (!disabled) {
                describe('And add to cart button is clicked', () => {
                    beforeAll(() => {
                        vi.useFakeTimers();
                        render(<AddToCart {...defaultProps} disabled={disabled} />);
                        fireEvent.click(screen.getByTestId('add-to-cart-button'));
                    });

                    afterAll(() => {
                        vi.useRealTimers();
                    });

                    it('Then snackbar should be open', () => {
                        expect(screen.getByTestId('add-to-cart-snackbar')).toBeInTheDocument();
                    });

                    it('Then snackbar should disappear after 2000ms', () => {
                        vi.advanceTimersByTime(2000);
                        expect(screen.queryByTestId('add-to-cart-snackbar')).not.toBeInTheDocument();
                    });
                });
            }
        })
    });

    [defaultProps.price, undefined].forEach((price) => {
        describe(`When the component is rendered with price=${price}`, () => {
            it(`Then it should ${price ? 'show' : 'not show'} the price`, () => {
                render(<AddToCart {...defaultProps} price={price} />);
                
                if (price) {
                    expect(screen.getByTestId('add-to-cart-price')).toBeInTheDocument();
                    expect(screen.getByTestId('add-to-cart-price')).toHaveTextContent('999.00 â‚¬');
                } else {
                    expect(screen.queryByTestId('add-to-cart-price')).not.toBeInTheDocument();
                }
            })
        })
    });
});