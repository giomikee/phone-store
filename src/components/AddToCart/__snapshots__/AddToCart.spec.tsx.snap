// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given AddToCart component > When the component is rendered > Then it should match the snapshot 1`] = `
<div>
  <div
    class="MuiStack-root add-to-cart css-15bx9vo-MuiStack-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorSuccess css-1l4a4zd-MuiButtonBase-root-MuiButton-root"
      data-testid="add-to-cart-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeLarge css-hev6se-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="AddShoppingCartIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2m10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2m-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25"
          />
        </svg>
      </span>
      Añadir al carrito
    </button>
    <h3
      class="MuiTypography-root MuiTypography-h5 add-to-cart__price css-5wi9gl-MuiTypography-root"
      data-testid="add-to-cart-price"
    >
      999.00 €
    </h3>
  </div>
</div>
`;
